Address autofocus issue #7711

mapitman opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
Remove the hard-coded `autoFocus` attributes from the auth-related jsx files. Replace with JavaScript to set the focus to the first `input` element in the auth modal.



### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- Use the magic "Fixes #1234" format, so the issues are -->
<!--- automatically closed when this PR is merged. -->

Fixes #7711


### How Has This Been Tested?
<!--- Please describe in detail how you manually tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

I tested it with using the swagger.json linked to in the issue. Also the automated test that was added to test the autoFocus still passes after this change.


### Screenshots (if appropriate):



## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

### My PR contains... 
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [x] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

There is an existing test for this.

mapitman commented on src/core/components/auth/auths.jsx:
```diff
@@ -17,6 +17,13 @@ export default class Auths extends React.Component {
   }
 
+  componentDidMount() {
+    var input = document.getElementsByTagName("input")[1]
```
I don't know that this is the most elegant way to handle this. If there is a more idiomatic method in React, let me know!

mapitman committed:
Address autofocus issue #7711

Find the first input element in the div with class name
modal-ux-content and set focus to it.

tim-lai commented:
@mapitman Thanks for the PR (and issue ticket). I do think that moving control of the prop to the top-level Auth component is a good starting point. However, I think there should be a way to provide a priority list of preferred components (keys?) to set an autofocus on first, and then skip for the rest of components if autofocus is set. I know you mentioned that you are not as familiar with React, so perhaps someone else is able to further develop this PR.

mapitman commented:
@tim-lai I'd be willing to take another go at this. Would you be willing to give me some guidance on how this would work? I'm not quite understanding the use case. Is it that someone may want to set focus to the second textbox in the modal for example?

tim-lai commented:
@mapitman  My interpretation of the earlier PR #6483 was that the PR simply added the autofocus to any input elements that could be present (per various individual use cases), without considering if those input elements could appear together. In the case(s) where multiple input elements are present, it seems reasonable that the autofocus should be on the first input element (this PR). So my request is that any proposed solution does NOT hardcode which input element to autofocus, but that it finds the first element rendered to which autofocus should be applied. Let me if know that helps clarify my request. Thanks!

mapitman commented:
@tim-lai Thanks for the reply. I've made a slight change because I noticed I was using using a class name specific to my use of Swagger UI. I've also rebased my changes on current master.

The code will now find the first input element in the modal div (found by using the class name `modal-ux-content`) and set focus to it. I have also removed all the `autoFocus` attributes from the inputs.

I believe this addresses your point above:
>So my request is that any proposed solution does NOT hardcode which input element to autofocus, but that it finds the first element rendered to which autofocus should be applied.

mapitman committed:
Remove semicolon

I didn't realize this project does not use semi-colons.

tim-lai commented on src/core/components/auth/auths.jsx:
```diff
@@ -17,6 +17,17 @@ export default class Auths extends React.Component {
   }
 
+  componentDidMount() {
+    var inputs = document.getElementsByClassName("modal-ux-content")[0].getElementsByTagName("input")
```
As a vanilla JS solution, this looks good to me. However, thinking about this more, we really should "React-ify" this to not directly change the DOM with the `.focus`. E.g. use "refs"

I'm thinking would could try adding a ref to `className="auth-container"` [here](https://github.com/swagger-api/swagger-ui/pull/7712/files#diff-e45dd3fb35ca515a5577879c3ac47d11ef521b2a8b00eadca99cecfb2baffd8dR85) that would contain multiple current refs. Most likely then also need to add creation of refs within the two maps that generate React element. e.g. something similar to `ref={el => inputRef.current[x] = el} `.


More info on refs [here](https://reactjs.org/docs/refs-and-the-dom.html), and we'll want to be compatible with the possible future use of React hooks.
